internal/providers/cloudstackMetadata/cloudstack.go

```diff
@@ -0,0 +1,181 @@
+// Copyright 2017 CoreOS, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package cloudstackMetadata
+
+import (
+	"fmt"
+	"path/filepath"
+	"strings"
+
+	"errors"
+	"github.com/coreos/coreos-metadata/internal/providers"
+	"io/ioutil"
+	"os"
+	"os/exec"
+	"syscall"
+)
+
+const (
+	diskByLabelPath               = "/dev/disk/by-label/"
+	configDriveMetadataPath       = "/cloudstack/metadata/"
+	configDriveMetadataMountPoint = "/media/ConfigDrive/cloudstack/metadata/"
+)
+
+func FetchMetadata() (providers.Metadata, error) {
+	m := providers.Metadata{}
+	m.Attributes = make(map[string]string)
+	mnt := ""
+
+	if !fileExists(configDriveMetadataMountPoint) {
+		err := error(nil)
+		mnt, err = mountConfigDrive("config")
+		if err != nil {
+			if mnt, err = mountConfigDrive("CONFIG"); err != nil {
+				return m, err
+			}
+		}
+		defer unmountConfigDrive(mnt)
+	} else {
+		mnt = configDriveMetadataMountPoint
+	}
+
+	if err := fetchAndSet("availability_zone.txt", mnt, "CLOUDSTACK_AVAILABILITY_ZONE", m.Attributes); err != nil {
+		return providers.Metadata{}, err
+	}
+
+	if err := fetchAndSet("instance_id.txt", mnt, "CLOUDSTACK_INSTANCE_ID", m.Attributes); err != nil {
+		return providers.Metadata{}, err
+	}
+
+	if err := fetchAndSet("service_offering.txt", mnt, "CLOUDSTACK_SERVICE_OFFERING", m.Attributes); err != nil {
+		return providers.Metadata{}, err
+	}
+
+	if err := fetchAndSet("cloud_identifier.txt", mnt, "CLOUDSTACK_CLOUD_IDENTIFIER", m.Attributes); err != nil {
+		return providers.Metadata{}, err
+	}
+
+	if err := fetchAndSet("local_hostname.txt", mnt, "CLOUDSTACK_LOCAL_HOSTNAME", m.Attributes); err != nil {
+		return providers.Metadata{}, err
+	}
+
+	if err := fetchAndSet("vm_id.txt", mnt, "CLOUDSTACK_VM_ID", m.Attributes); err != nil {
+		return providers.Metadata{}, err
+	}
+
+	keys, err := fetchKeys(mnt)
+	if err != nil {
+		return providers.Metadata{}, err
+	}
+	m.SshKeys = keys
+	return m, nil
+}
+
+func fileExists(path string) bool {
+	_, err := os.Stat(path)
+	return (err == nil)
+}
+
+func labelExists(label string) bool {
+	_, err := getPath(label)
+	return (err == nil)
+}
+
+func getPath(label string) (string, error) {
+	path := diskByLabelPath + label
+
+	if fileExists(path) {
+		return path, nil
+	}
+
+	out, err := exec.Command("blkid", "-l", "-t", "LABEL="+label, "-o", "device").Output()
```

Comment:
This step doesn't seem necessary since udev will make sure `/dev/disk/by/label/` is populated.